proposed/2021/MachineReadableOutput.md

```diff
@@ -0,0 +1,797 @@
+   > Text2Xml.Lib         1.1.2                 1.1.2
+
+   [net5.0]:
+   Top-level Package      Requested             Resolved
+   > NuGet.Commands       4.8.0-preview3.5278   4.8.0-preview3.5278
+   > Text2Xml.Lib         1.1.2                 1.1.2
+```
+
+### dotnet list package --json
+
+```json
+{
+  "version": 1,
+  "MyProjectA": [
+    {
+      "framework": "netcoreapp3.1",
+      "topLevelPackages": [
+        {
+          "id": "Microsoft.Extensions.Primitives",
+          "requestedVersion": "[1.0.0, 5.0.0]",
+          "resolvedVersion": "1.0.0"
+        },
+        {
+          "id": "NuGet.Commands",
+          "requestedVersion": "4.8.0-preview3.5278",
+          "resolvedVersion": "4.8.0-preview3.5278"
+        },
+        {
+          "id": "Text2Xml.Lib",
+          "requestedVersion": "[1.1.2, 2.0.0)",
+          "resolvedVersion": "1.1.2"
+        }
+      ]
+    }
+  ],
+  "MyProjectB": [
+    {
+      "framework": "netcoreapp3.1",
+      "topLevelPackages": [
+        {
+          "id": "NuGet.Commands",
+          "requestedVersion": "4.8.0-preview3.5278",
+          "resolvedVersion": "4.8.0-preview3.5278"
+        },
+        {
+          "id": "Text2Xml.Lib",
+          "requestedVersion": "1.1.2",
+          "resolvedVersion": "1.1.2"
+        }
+      ]
+    },
+    {
+      "framework": "net5.0",
+      "topLevelPackages": [ 
+        {
+          "id": "NuGet.Commands",
+          "requestedVersion": "4.8.0-preview3.5278",
+          "resolvedVersion": "4.8.0-preview3.5278"
+        },
+        {
+          "id": "Text2Xml.Lib",
+          "requestedVersion": "1.1.2",
+          "resolvedVersion": "1.1.2"
+        }
+      ]
+    }
+  ]
+}
+```
+
+### dotnet list package --outdated
+
+```dotnetcli
+
+The following sources were used:
+   https://api.nuget.org/v3/index.json
+   https://apidev.nugettest.org/v3-index/index.json
+
+Project `MyProjectA` has the following updates to its packages
+   [netcoreapp3.1]:
+   Top-level Package                      Requested             Resolved              Latest
+   > Microsoft.Extensions.Primitives      [1.0.0, 5.0.0]        1.0.0                 6.0.0
+   > NuGet.Commands                       4.8.0-preview3.5278   4.8.0-preview3.5278   6.0.0
+   > Text2Xml.Lib                         [1.1.2, 2.0.0)        1.1.2                 1.1.4
+
+Project `MyProjectB` has the following updates to its packages
+   [netcoreapp3.1]:
+   Top-level Package      Requested             Resolved              Latest
+   > NuGet.Commands       4.8.0-preview3.5278   4.8.0-preview3.5278   6.0.0
+   > Text2Xml.Lib         1.1.2                 1.1.2                 1.1.4
+
+   [net5.0]:
+   Top-level Package      Requested             Resolved              Latest
+   > NuGet.Commands       4.8.0-preview3.5278   4.8.0-preview3.5278   6.0.0
+   > Text2Xml.Lib         1.1.2                 1.1.2                 1.1.4
+```
+
+### dotnet list package --outdated --json
```

Comment:
I agree, I would expect `--outdated` to just show more metadata like
```json
"outdated": true
```

Having to run the command twice seems bad, the point is to give something readable and the user can process the output in whatever manner they want.

Reply:
And the same with `--deprecated`

Reply:
Do you have a proposal of what the ideal experience should look like? I agree that most customers who use the json output will probably want both, but we also want the output to meet their expectations based on existing behavior.

Would be weird to have `dotnet list package --outdated --json` also output vulnerable packages if they only expect outdated packages.

Similarly, most customers will probably also want transitive dependencies to create an SBOM. 

Personally, I think the dotnet cli needs an all up `dotnet list package --all` that shows all deprecated, vulnerable, and outdated top level and transitive packages.

Altneraitvely the behavior of `--outdated` and `--deprecated` could be additive rather than exclusive. That adds to the scope of this work though.

Reply:
```json
 {
          "id": "Microsoft.Extensions.Primitives",
          "requestedVersion": "[1.0.0, 5.0.0]",
          "resolvedVersion": "1.0.0",
          "latestVersion": "6.0.0",
          "deprecated": "true",
          "vulnerable": "details?",
        },
```

Should the existence of `latestVersion` imply outdated? 

There is obv a world where the installed version is *not* the latest available though, so maybe that warrants the existence of both.

Reply:
fwiw, I actually think that we should design one format for *all* combinations, however you *should* be able to scope the output if you wanted to.

Reply:
Please refer to this comment: https://github.com/NuGet/Home/issues/7752#issuecomment-740206375

> Human-readable output is first and foremost. This should be covered with the current list package command, however we may have some improvements coming to the command to make this better to glance at.

> Machine-readable output is second to make output easier to handle, especially complex lists this command can produce (especially with transitive dependencies included)

The default format should be human-readable. The specified format should be machine-readable. 

We should reach out to Chet Husk to help us understand what would scale with other `dotnet CLI` commands before leaning on one side.

@chgill-MSFT That is something I originally proposed but sadly we never got to it. We will have a similar experience in https://github.com/NuGet/Home/blob/dotnet-audit/proposed/2021/DotNetAudit.md

Reply:
Won't fix, out of scope from MVP. **_The behavior of the command is separate from the json output._** I created tracking issue. https://github.com/NuGet/Home/issues/11551